src/SignalR/common/Shared/ClientResultsManager.cs

```diff
@@ -0,0 +1,193 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.AspNetCore.SignalR.Protocol;
+
+namespace Microsoft.AspNetCore.SignalR.Internal;
+
+// Common type used by our HubLifetimeManager implementations to manage client results.
+// Handles cancellation, cleanup, and completion, so any bugs or improvements can be made in a single place
+internal class ClientResultsManager : IInvocationBinder
+{
+    private readonly ConcurrentDictionary<string, (Type Type, string ConnectionId, object Tcs, Action<object, CompletionMessage> Complete)> _pendingInvocations = new();
+
+    public Task<T> AddInvocation<T>(string connectionId, string invocationId, CancellationToken cancellationToken)
+    {
+        var tcs = new TaskCompletionSourceWithCancellation<T>(this, connectionId, invocationId, cancellationToken);
+        var result = _pendingInvocations.TryAdd(invocationId, (typeof(T), connectionId, tcs, static (state, completionMessage) =>
+        {
+            var tcs = (TaskCompletionSourceWithCancellation<T>)state;
+            if (completionMessage.HasResult)
+            {
+                tcs.SetResult((T)completionMessage.Result);
+            }
+            else
+            {
+                tcs.SetException(new Exception(completionMessage.Error));
+            }
+        }
+        ));
+        Debug.Assert(result);
```

Comment:
Well lots of things can happen if there is a misbehaving server :)

We always generate a unique ID for the invocation so this will never be an issue for our servers right now. If we do collide then it's a big problem.